File: 剑指offer/3.数组中重复的数字/set.go
package main

import "fmt"

type void struct{}

func findDuplicate(arr []int) *int {
	// more memory efficient than map[int]bool
	set := make(map[int]void)

	for _, num := range arr {
		_, ok := set[num]
		if ok {
			return &num
		}
		set[num] = void{}
	}

	return nil
}

func main() {
	input := []int{2, 3, 1, 0, 2, 5, 3}
	output := findDuplicate(input)
	if output != nil {
		fmt.Println(*output)
	} else {
		fmt.Println("nil")
	}
}

File: 剑指offer/3.数组中重复的数字/set.ts
/**
 * 方法一：利用一个set将已经找到的数字记录下来
 */
function findDuplicate(arr: number[]): number | undefined {
  const set = new Set<number>();

  for (let i = 0; i < arr.length; i++) {
    if (set.has(arr[i])) {
      return arr[i];
    }
    set.add(arr[i]);
  }

  return undefined;
}

const arr = [2, 3, 1, 0, 2, 5, 3];

const output = findDuplicate(arr);
console.log(output);
